stakingModel/contracts/InterestDistribution.sol

```diff
@@ -0,0 +1,211 @@
+      * @param _staker                 Staker's address
+      * @param _stake                  Amount of stake
+      * @param _donationPer            Percentage will to donate.
+      * @param _dailyIntrest            Withdrawn interest by individual staker so far.
+      * @param _grandTotalStaked        Average yielding rate per token
+      *
+    */
+    function stakeCalculation(
+      InterestData storage _interestData, 
+      YieldData storage _yieldData,
+      address _staker,
+      uint256 _stake, 
+      uint256 _donationPer,
+      uint256 _dailyIntrest,
+      uint256 _grandTotalStaked
+      ) 
+    internal 
+    {
+      _interestData.grandTotalStaked = _interestData.grandTotalStaked.add(_stake);
+      uint currentStake = _interestData.stakers[_staker].stakedToken;
+      _interestData.stakers[_staker].stakedToken = currentStake.add(_stake);
+      _interestData.stakers[_staker].weightedStake = (_interestData.stakers[_staker].weightedStake.mul(currentStake).add(_stake.mul(uint(100).sub(_donationPer)))).div(_interestData.stakers[_staker].stakedToken);
+      _interestData.stakers[_staker].lastStake = block.number;
+      uint _accumulatedYieldPerToken = getAccumulatedYieldPerToken(_dailyIntrest, _grandTotalStaked);
+      _yieldData.avgYieldRatePerToken[_staker] = _yieldData.avgYieldRatePerToken[_staker].add(getAvgYieldRatePerToken(_accumulatedYieldPerToken, _stake, _donationPer, _interestData.stakers[_staker].stakedToken));
+    }
+
+    /**
+      * @dev Updates InterestData and Staker data while withdrawing stake.
+      * 
+      * @param _interestData           Interest data
+      * @param _staker                 Staker address
+      * @param _amount                 Amount of stake to withdraw
+      *
+    */
+    function withdrawStakeAndInterest(
+      InterestData storage _interestData, 
+      address _staker, 
+      uint256 _amount
+      ) 
+    internal 
+    {
+      _interestData.grandTotalStaked = _interestData.grandTotalStaked.sub(_amount);
+      _interestData.stakers[_staker].stakedToken = _interestData.stakers[_staker].stakedToken.sub(_amount);
+    }
+
+    /**
+      * @dev Updates Accumulated Yield Per Token.
+      * 
+      * @param _yieldData                   Yield data
+      * @param _accumulatedYieldPerToken    Total yielding amount per token
+      *
+    */
+    function addAccumulatedYieldPerToken(
+      YieldData storage _yieldData, 
+      uint _accumulatedYieldPerToken
+      ) 
+    internal 
+    {
+      _yieldData.accumulatedYieldPerToken = _yieldData.accumulatedYieldPerToken.add(_accumulatedYieldPerToken);
+    }
+
+    /**
+      * @dev Updates interestAccrued.
+      * 
+      * @param _interestData           Interest data
+      * @param _iterest                Newly accrued intrest
+      *
+    */
+    function addInterest(InterestData storage _interestData, uint256 _iterest) internal {
+      _interestData.interestAccrued.add(_iterest);
```

Comment:
Yes, it will make more sense to update both at same time. Will remove updateAccumulatedYieldPerToken() and include its updating in addInterest() itself.